### PR TITLE
windows ddupl grabber downscaling

### PR DESCRIPTION
**Pull request by zomfg**
I added the ability to downscale the final texture before copying the data.

The caveat for now is that the division is limited to powers of 2: 0 would divide width and height by 1, 1 by 2, 2 by 4 etc.
On my setup the optimal value seems to be 3 (width and height divided 8), anything higher looks counter productive CPU/GPU usage-wise, so I hardcoded the 3 for now.

Ideally I would like it be a percentage, but I have literally no idea what I'm doing with D3D11, so if you know how to achieve that, that would be cool. Or if you know why it's getting worse with 4+ values..
The reason why I'd like a percentage is because on a new version of macos grabber I'm working on it is possible to do so and if all grabbers are able to do it, it'd be easier to make it available in the settings.

Now, to support this, I added `scale` to `GrabbedScreen` which is set by the grabber and then used to scale down the widget `QRect` before color math.
Slightly off-topic: I also added `bytesPerRow` to `GrabbedScreen` to be able to override (width * 4) value passed to `Calculations::calculateAvgColor` because not all grabbing methods will produce `bytesPerRow == (width * bytesPerPixel)` due to padding and what not.

Performance wise, going from `0` setting (no scaling) to `3`, my CPU usage goes down by half, and GPU divided by 3-4 :)

### CODE REVIEW

**Comment by zomfg** [comment]
At this point performance cost of Prismatik is reduced by more than half while running 3DMark Time Spy

|Prismatik| pts |
|-|-|
| off  | ~4500  |
| 16ms / 2^0 scaling  | ~4200  |
| 16ms / 2^3 scaling  | ~4390  |

**Comment by psieg** [comment]
I think it is getting worse with higher number because the GPU will always create all MIP maps.
I'll try to look for a method to scale down using DXGI.

**Comment by psieg** [comment]
Well since Prismatik does nothing other than taking average colors, which is what downscaling will do I'm not too surprised. I'l try to dig in a little, but no timeline...

**Comment by zomfg** [comment]
Well, worst case, this PR is usable as is, so you can dig around any time :)

**Comment by psieg** [comment]
ok to test

**Comment by zomfg** [comment]
couple of other projects using same method stop at 3 as well

File: Software/grab/GrabberBase.cpp [hunk 1]
```diff
@@ -165,9 +165,19 @@ void GrabberBase::grab()
 			// Convert coordinates from "Main" desktop coord-system to capture-monitor coord-system
 			QRect preparedRect = clippedRect.translated(-monitorRect.x(), -monitorRect.y());
 
+			// grabbed screen was scaled => scale the widget
+			if (grabbedScreen->scale != 1.0)
+				preparedRect.setCoords(
+					std::floor(grabbedScreen->scale * preparedRect.left()),
+					std::floor(grabbedScreen->scale * preparedRect.top()),
+					std::floor(grabbedScreen->scale * preparedRect.right()),
+					std::floor(grabbedScreen->scale * preparedRect.bottom())
```
[the remark this answers is not in the document]
Comment (zomfg) [reply Software/grab/GrabberBase.cpp]: Indeed. Bottom-right it is.

**Comment by zomfg** [comment]
There's one other thing I tried unsuccessfully.
The [documentation](https://docs.microsoft.com/en-us/windows/desktop/api/dxgi1_2/nf-dxgi1_2-idxgioutputduplication-releaseframe#remarks) says:
> For performance reasons, we recommend that you release the frame just before you call the IDXGIOutputDuplication::AcquireNextFrame method to acquire the next frame. When the client does not own the frame, the operating system copies all desktop updates to the surface. This can result in wasted GPU cycles if the operating system updates the same region for each frame that occurs. When the client acquires the frame, the client is aware of only the final update to this region; therefore, any overlapping updates during previous frames are wasted. When the client acquires a frame, the client owns the surface; therefore, the operating system can track only the updated regions and cannot copy desktop updates to the surface. Because of this behavior, we recommend that you minimize the time between the call to release the current frame and the call to acquire the next frame.

I tried releasing right before `Acquire` but it cuts the FPS in half. Maybe this is too fast and it needs time for a new frame to come in. I didn't look deeper.

**Comment by psieg** [comment]
Yeah not sure. Performance wise, it would be better to release just before acquire. As you discovered this means that the Acquire function will actually wait for the next screen update (i.e. frame). This means the grabbing takes more time, thus the actual framerate drops. However this can easily be compensated by reducing the sleep duration.

It just means that the "theoretical FPS" are less realistic, because that computation assumes zero time for grab. If we used a different formula, that UI weirdness could be fixed while keeping the performance impact - it might just be counter-intuitive for advanced users.

**Comment by psieg** [comment]
Probably it makes the most sense to not show the max FPS because it is very hard to determine correctly and will depend on hardware, software timers etc. It won't be as satisfying to use/watch but lower performance overhead is better...

File: Software/grab/GrabberBase.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (zomfg) [reply Software/grab/GrabberBase.cpp]: ...or not. It goes 1 line to far down now and segfaults on bottom right widget.
Switching `x/y/width/height` to `left/top/right/bottom` in `calculations.cpp` works but `bottom/right` do `-1`, which would be the same as `floor`ing before... or adding yet another `preparedRect` validation/control in `GrabberBase::grab()`, but might as well keep it simple with floor, imo.
Or I'm too tired and missing the obvious somewhere.

relevant config for 3440x1440
```ini
[LED_1]
Position=@Point(2924 1338)
Size=@Size(516 102)

[LED_2]
Position=@Point(2924 1236)
Size=@Size(516 102)
```

[the remark this answers is not in the document]
Comment (zomfg) [reply Software/grab/GrabberBase.cpp]: or maybe something like this:
```cpp
std::min(std::ceil(grabbedScreen->scale * preparedRect.right()),  std::floor(grabbedScreen->scale * monitorRect.right())),
std::min(std::ceil(grabbedScreen->scale * preparedRect.bottom()), std::floor(grabbedScreen->scale * monitorRect.bottom()))
```

[the remark this answers is not in the document]
Comment (psieg) [reply Software/grab/GrabberBase.cpp]: I switched ceil and floor, i.e. opted for the smaller rect. Unfortunately the dimensions of the grabbed texture are not available at that point, and rounding the bound would result in the same error.

[the remark this answers is not in the document]
Comment (zomfg) [reply Software/grab/GrabberBase.cpp]: top-left is risky too? isn't it naturally limited to 0? or it's for the margin between widgets?

[the remark this answers is not in the document]
Comment (psieg) [reply Software/grab/GrabberBase.cpp]: top left is safe for overflow, but as mentioned in the original comment I'd like it to be consistent - both rounded to the inside or both to the outside.

**Comment by zomfg** [comment]
I agree about max FPS inaccuracy (and yes, "1000 FPS" looks silly) but I find it handy (it's a more straight forward value) as it works ok in the lower range so I'd vote to put what you just said basically in a "?" disclaimer right beside it to manage user expectations.. or, probably better, show it only when "Expert mode" is checked for example.

**Comment by psieg** [comment]
For posterity, this is responsible for crashes: There are errors with memory management: the `*Ptr` types are com smart pointers that don't need to be released. Setting to `nullptr` and destruction releases them.

I wasn't thinking of it when reviewing. I have no explanation why it works on our machines. I crashes on my laptop but works fine on my desktop. No idea.

See e9e6e2f06b953aee38e6de1db915b06414c848c

**Comment by zomfg** [comment]
Ooh, noted.
So this was #240 ?
Do you have the exact same setup on the laptop? maybe slightly different DLL version of something

**Comment by psieg** [comment]
Obviously not exact same. Both have VS, usually that's the main difference between dev and user machines. Probably it has an older runtime version, but I don't see why that would matter.

**Pull request merged**